src/nacl/bindings/crypto_generichash.py

```diff
@@ -112,3 +119,101 @@ def generichash_blake2b_salt_personal(data,
+    :type salt: bytes
+    :param person: must be at most
+                   :py:data:`.crypto_generichash_PERSONALBYTES` long:
+                   will be zero-padded if needed
+    :type person: bytes
+    :param digest_size: must be at most
+                        :py:data:`.crypto_generichash_BYTES_MAX`;
+                        the default digest size is
+                        :py:data:`.crypto_generichash_BYTES`
+    :type digest_size: int
+    :return: an initizialized state buffer
+    :rtype: bytes
+    """
+
+    _checkparams(digest_size, key, salt, person)
+
+    statebuf = ffi.new("unsigned char[]", crypto_generichash_STATEBYTES)
+    _state = ffi.cast("struct crypto_generichash_blake2b_state *", statebuf)
+
+    # both _salt and _personal must be zero-padded to the correct length
+    _salt = ffi.new("unsigned char []", crypto_generichash_SALTBYTES)
+    _person = ffi.new("unsigned char []", crypto_generichash_PERSONALBYTES)
+
+    ffi.memmove(_salt, salt, len(salt))
+    ffi.memmove(_person, person, len(person))
+
+    rc = lib.crypto_generichash_blake2b_init_salt_personal(_state,
+                                                           key, len(key),
+                                                           digest_size,
+                                                           _salt, _person)
+    ensure(rc == 0, 'Unexpected failure',
+           raising=exc.RuntimeError)
+
+    return statebuf
+
+
+def generichash_blake2b_update(statebuf, data):
+    """Update the blake2b hash state
+
+    :param statebuf: an initialized blake2b state buffer as returned from
+                     :py:func:`.crypto_generichash_blake2b_init`
+    :type name: bytes
+    :param data:
+    :type data: bytes
+    """
+
+    ensure(isinstance(data, bytes),
+           'Input data must be a bytes sequence',
+           raising=exc.TypeError)
+
+    _state = ffi.cast("struct crypto_generichash_blake2b_state *", statebuf)
+
+    rc = lib.crypto_generichash_blake2b_update(_state, data, len(data))
+    ensure(rc == 0, 'Unexpected failure',
+           raising=exc.RuntimeError)
+
+
+def generichash_blake2b_final(statebuf, digest_size):
+    """Finalize the blake2b hash state and return the digest.
+
+    :param statebuf:
+    :type statebuf: bytes
+    :param digest_size:
+    :type digest_size: int
+    :return: the blake2 digest of the passed-in data stream
+    :rtype: bytes
+    """
+
+    _digest = ffi.new("unsigned char[]", crypto_generichash_BYTES_MAX)
+    _state = ffi.cast("struct crypto_generichash_blake2b_state *", statebuf)
+    rc = lib.crypto_generichash_blake2b_final(_state, _digest, digest_size)
+
+    ensure(rc == 0, 'Unexpected failure',
+           raising=exc.RuntimeError)
+    return ffi.buffer(_digest, digest_size)[:]
+
+
+def generichash_blake2b_state_copy(statebuf):
+    """Return a copy of the given blake2b hash state"""
+
+    _statebuf = ffi.new("unsigned char[]", crypto_generichash_STATEBYTES)
```

Comment:
I just tried following your suggestion, but I quickly get a segmentation fault, which I don't fully understand whether to blame to a lifecycle problem with the underlying buffer, or to the different sizes of the buffer itself and the embedded struct.

The only way I've found to avoid the segmentation fault, and hiding the cast has been defining an helper class to keep the state as in:
```
class _blake2b_state(object):
    """Utility class to keep around the blake2b state"""

    def __init__(self, buf = None):
        self.statebuf = ffi.new("unsigned char[]",
                                crypto_generichash_STATEBYTES)
        if buf is not None:
            ffi.memmove(self.statebuf, buf, crypto_generichash_STATEBYTES)

    @property
    def b2state(self):
        return ffi.cast("struct crypto_generichash_blake2b_state *",
                        self.statebuf)
```

and express the operations in terms of that partially opaque class like in:

```
def generichash_blake2b_update(state, data):
    """ [ . . . ] """
    rc = lib.crypto_generichash_blake2b_update(state.b2state, data, len(data))

```
but I'm really unsure about the worth of such an indirection.

I've added the test commit to a branch in my forked repository; to browse the changes, you can open

https://github.com/lmctv/pynacl/compare/add-generichash-iuf...lmctv:add-generichash-iuf-hide-casts?expand=1